### DISCUSSION TITLE
Proposal : diagnostic when caller provides a value for a parameter marked with a Caller Info attribute

### DISCUSSION THREAD

**Discussion by nalka0** [dotnet/csharplang#6741] [Language Ideas]
The way I see it : adding a constructor to Caller Info attributes (such as [CallerMemberName](https://learn.microsoft.com/fr-fr/dotnet/api/system.runtime.compilerservices.callermembernameattribute?view=net-7.0)) with a (probably?) [DiagnosticSeverity](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnosticseverity?view=roslyn-dotnet-4.3.0) as parameter.
When the provided DiagnosticSeverity is different than Hidden, if a caller provides a value for the parameter that was marked with the Caller Info attribute they get a CSXXXX diagnostic with the severity passed to the attribute's constructor.

For example, if I declare a method like this :
```cs
public void DoSomething([CallerMemberName(DiagnosticSeverity.Warning)] string callerName = null)
{
    // Do something with caller's name
}
```
then calling it with `DoSomething("Some string");` would give me a warning.

The point of this would be either preventing accidentally providing a value for the parameter or even forbidding it if you use Error as severity. 

In fact this is pretty similar to ObsoleteAttribute's constructor that uses a bool to determine if the caller should recieve a warning or an error.

**Comment by HaloFour** [comment]
You can write an analyzer which emits a warning if someone manually passes a value to one of these parameters.

**Reply by nalka0** [reply]
I have already thought about doing this but it wouldn't let me
1) Choose when I want a warning and when I don't
2) Choose the severity

Also the point of having it built-in is that everyone wouldn't need to write their own analyser

**Reply by HaloFour** [reply]
> I have already thought about doing this but it wouldn't let me
> 
> 1. Choose when I want a warning and when I don't
> 2. Choose the severity

You could accomplish both through a second attribute.  What your analyzer couldn't do is to add a new constructor parameter to the existing attribute.

> Also the point of having it built-in is that everyone wouldn't need to write their own analyser

No, but it's something that you could accomplish today.  Warning on the use of an existing language feature based on your own rules is exactly why analyzers were added to the compiler.

**Reply by nalka0** [reply]
> No, but it's something that you could accomplish today. Warning on the use of an existing language feature based on your own rules is exactly why analyzers were added to the compiler.

Then we could also argue wheter ObsoleteAttribute should be provided by the language or people should write their own analysers to warn about obsolete things but I didn't come here to talk about this (I'm not saying that it's a question that doesn't matter).

**Reply by HaloFour** [reply]
> Then we could also argue wheter ObsoleteAttribute should be provided by the language or people should write their own analysers to warn about obsolete things but I didn't come here to talk about this (not saying that it's a question that doesn't matter).

We could, although that existed long before analyzers were added to the compiler, and I would suggest that the team would agree that marking types and members as obsolete is more universally useful.  Not to mention, it's used by languages other than C#/VB.NET.

**Reply by CyrusNajmabadi** [reply]
Yes, we could. If we had analyzers first, I would certainly argue that that's how it should have worked :-)

**Reply by nalka0** [reply]
Just being curious, since we have IL weaving do you both also consider people should write their own weavers instead of having the caller info attributes?

**Reply by HaloFour** [reply]
Weavers don't exist as a part of the compiler, but their existence and ease of use would be considered as a means to solve a particular problem over a language feature that does the same.  It depends on a lot of factors, and ROI weighs heavily into that.  Having to modify the language to solve a particular problem is an extremely expensive solution with likely the longest lead time.

**Comment by sab39** [comment]
You'd want to make sure there was an exception for cases like:

```cs
void Foo(FooParam fooParam, [CallerMemberName] string caller = null) { ... }

void Foo([CallerMemberName] string caller = null) => Foo(getDefaultFooParam(), caller);
```

(and yeah, this would be best as an analyzer, but it'd be nice if it were one of the analyzers shipped by default. I guess that would be a request to file in a different git project though?)

**Reply by nalka0** [reply]
> I guess that would be a request to file in a different git project though?

I chose to post here instead of [dotnet/roslyn](https://github.com/dotnet/roslyn/) because I saw [their #351](https://github.com/dotnet/roslyn/issues/351) (similar topic) was moved here after some time

**Reply by 333fred** [reply]
Such an analyzer would be best requested on the roslyn-analyzers repo.

**Reply by jnm2** [reply]
https://www.nuget.org/packages/PropertyChangedAnalyzers/ does this checking.

**Reply by nalka0** [reply]
> https://www.nuget.org/packages/PropertyChangedAnalyzers/ does this checking.

It does only for CallerMemberName, not other attributes
